restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -84,6 +95,96 @@ static Object translateCanonicalKeyToAlternativeKeyIfNeeded(Object canonicalKey,
+        return fillInDefaultOnRecord((RecordDataSchema) schema, (DataMap) dataWithoutDefault);
+      case TYPEREF:
+        return fillInDefaultOnTyperef((TyperefDataSchema) schema, dataWithoutDefault);
+      case MAP:
+        return fillInDefaultOnMap((MapDataSchema) schema, (DataMap) dataWithoutDefault);
+      case UNION:
+        return fillInDefaultOnUnion((UnionDataSchema) schema, (DataMap) dataWithoutDefault);
+      case ARRAY:
+        return fillInDefaultOnArray((ArrayDataSchema) schema, (DataList) dataWithoutDefault);
+      default:
+        return dataWithoutDefault;
+    }
+  }
+
+  public static DataMap fillInDefaultOnRecord(RecordDataSchema schema, DataMap dataMap)
+  {
+    DataMap dataWithDefault = new DataMap(DataMapBuilder.getOptimumHashMapCapacityFromSize(dataMap.size()));
+    dataWithDefault.putAll(dataMap);
+    for (RecordDataSchema.Field field : schema.getFields())
+    {
+      if (dataMap.containsKey(field.getName()) || field.getDefault() != null)
+      {
+        Object fieldData = dataMap.containsKey(field.getName()) ? dataMap.get(field.getName()) : field.getDefault();
+        dataWithDefault.put(field.getName(), fillInDataDefault(field.getType(), fieldData));
+      }
+    }
+    return dataWithDefault;
+  }
+
+  public static DataMap fillInDefaultOnMap(MapDataSchema schema, DataMap dataMap)
+  {
+    DataSchema valueSchema = schema.getValues();
+    DataMap dataWithDefault = new DataMap(dataMap);
+    for (Map.Entry<String, Object> entry : dataWithDefault.entrySet())
+    {
+      dataWithDefault.put(entry.getKey(), fillInDataDefault(valueSchema, entry.getValue()));
+    }
+    return dataWithDefault;
+  }
+
+  public static DataList fillInDefaultOnArray(ArrayDataSchema schema, DataList dataList)
+  {
+    DataSchema itemDataSchema = schema.getItems();
+    DataList dataListWithDefault = new DataList();
+    for (Object o : dataList)
+    {
+      dataListWithDefault.add(fillInDataDefault(itemDataSchema, o));
+    }
+    return dataListWithDefault;
+  }
+
+  public static DataMap fillInDefaultOnUnion(UnionDataSchema schema, DataMap dataMap)
+  {
+    DataMap dataWithDefault = new DataMap(dataMap);
+    if (dataWithDefault.size() == 1)
+    {
+      for (Map.Entry<String, Object> entry: dataWithDefault.entrySet())
+      {
+        String memberTypeKey = entry.getKey();
+        DataSchema memberDataSchema = schema.getTypeByMemberKey(memberTypeKey);
+        if (memberDataSchema == null)
+        {
+          return dataWithDefault;
+        }
+        dataWithDefault.put(memberTypeKey, fillInDataDefault(memberDataSchema, entry.getValue()));
+      }
+    }
+    return dataWithDefault;
+  }
+
+  public static Object fillInDefaultOnTyperef(TyperefDataSchema typerefDataSchema, Object data)
+  {
+    DataSchema dataSchema = typerefDataSchema.getDereferencedDataSchema();
+    return fillInDataDefault(dataSchema, data);
+  }
+
+  public static DataMap fillInDefaultValues(DataSchema dataSchema, DataMap dataMap)
```

Comment:
You don't need this method. Just make fillInDataDefault public and use that.

Reply:
Method consolidated